src/v/storage/spill_key_index.cc

```diff
@@ -108,47 +108,54 @@ ss::future<> spill_key_index::add_key(compaction_key b, value_type v) {
     if (
       (take_result.checkpoint_hint && expected_size > min_index_size)
       || expected_size >= _max_mem) {
-        f = ss::do_until(
-          [this, entry_size, min_index_size] {
-              size_t total_mem = idx_mem_usage() + _keys_mem_usage + entry_size;
-
-              // Instance-local capacity check
-              bool local_ok = total_mem < _max_mem;
-
-              // Shard-wide capacity check
-              bool global_ok = _resources.compaction_index_bytes_available()
-                               || total_mem < min_index_size;
-
-              // Stop condition: none of our size thresholds must be violated
-              return _midx.empty() || (local_ok && global_ok);
-          },
-          [this] {
-              /**
-               * Evict first entry, we use hash function that guarante good
-               * randomness so evicting first entry is actually evicting a
-               * pseudo random elemnent
-               */
-              auto node = _midx.extract(_midx.begin());
-
-              return ss::do_with(
-                node.key(),
-                node.mapped(),
-                [this](const bytes& k, value_type o) {
-                    release_entry_memory(k);
-                    return spill(compacted_index::entry_type::key, k, o);
-                });
-          });
+        auto stop_cond = [this, entry_size, min_index_size] {
+            size_t total_mem = idx_mem_usage() + _keys_mem_usage + entry_size;
+
+            // Instance-local capacity check
+            bool local_ok = total_mem < _max_mem;
+
+            // Shard-wide capacity check
+            bool global_ok = _resources.compaction_index_bytes_available()
+                             || total_mem < min_index_size;
+
+            // Stop condition: none of our size thresholds must be violated
+            return _midx.empty() || (local_ok && global_ok);
+        };
+
+        auto iter = _midx.begin();
+        while (!stop_cond()) {
+            /**
+             * Evict first entry, we use hash function that guarante good
+             * randomness so evicting first entry is actually evicting a
+             * pseudo random elemnent
+             */
+            auto to_extract = iter;
+            iter++;
+            auto node = _midx.extract(to_extract);
+            auto capacity_before = _midx.capacity();
+
+            release_entry_memory(node.key());
+            co_await spill(
+              compacted_index::entry_type::key, node.key(), node.mapped());
+
+            if (capacity_before < _midx.capacity()) {
+                // Iterators into _midx only ever get invalidated by the insert
+                // below if a resize and rehash happens (from concurrent
+                // coroutines). Hence we can use capacity to check whether we
+                // suspended and the iterator got invalidated and we need to
+                // requery.
```

Comment:
This is a big change that looks small. These types of careful invalidation checks are something to avoid if at all possible I think. I recall a recent one in one of the metadata caches that led to UAF that took ages to track down. It is difficult to test this.

Are there any restrictions on what methods are called on spill key index concurrently? If not, I see clear() being called on the map, which invalidates all iterators (but reduces capacity) and the extract call below as as well as other extract calls which may erase the element `iter` points to which also invalidates its iterator.

Reply:
> This is a big change that looks small. These types of careful invalidation checks are something to avoid if at all possible I think.

Yep agree.

> clear() being called on the map

The clear is guarded by the gate.

> and the extract call below

Yeah that's why we increment the iterator first before extracting it.

> as well as other extract calls

Yeah this seems to be an issue. It's a bit sad because the map could handle this case gracefully as the iterator still points into a valid map but I guess we do need a generation counter after all.

Reply:
Yeah this is getting more complicated than I wanted it to be, likely going to close depending on the other comment as the benefits are hard to measure so it's a bit questionable.

Reply:
> Yeah that's why we increment the iterator first before extracting it.

Yes, for "this fiber" only that seems OK but is there something stopping > 1 fibers from calling `add_key` concurrently? 

So if the index has {a, b} and fiber 0 comes and increments its iterator to point to b and extracts a, suspends then fiber 1 comes in and extracts b, invalidating fiber 0's iterator w/o changing the capacity? 

Of course the possible execution paths depend on the allowed semantics around calling spill key index methods, currently undocumented.

What if we just extract all the keys first, then loop over all of them and spill them? I guess the stopping condition would have to be somewhat different because currently we rely on `release_entry_memory` to be called once per extraction to update the memory usage estimate.

Other option is just to maintain a separate `vector`-ish of `compaction_key`s in `_midx`: you can select randomly from this vector easily and erase that element from both. This solves the bias problem and probably ends up cheaper anyway because right now even the occasional `begin()` call is going to be very expensive over time as the map gets highly warped (all elements condensed on the right hand side of the bucket space). Of course, we won't want to duplicate the key itself.